cpp/src/parquet/geometry_statistics.cc

```diff
@@ -0,0 +1,278 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/geometry_statistics.h"
+#include <memory>
+
+#include "arrow/array.h"
+#include "arrow/type.h"
+#include "arrow/util/bit_run_reader.h"
+#include "arrow/util/logging.h"
+#include "parquet/geometry_util_internal.h"
+
+using arrow::util::SafeLoad;
+
+namespace parquet {
+
+class GeospatialStatisticsImpl {
+ public:
+  GeospatialStatisticsImpl() = default;
+  GeospatialStatisticsImpl(const GeospatialStatisticsImpl&) = default;
+
+  bool Equals(const GeospatialStatisticsImpl& other) const {
+    if (is_valid_ != other.is_valid_) {
+      return false;
+    }
+
+    if (!is_valid_ && !other.is_valid_) {
+      return true;
+    }
+
+    auto geospatial_types = bounder_.GeometryTypes();
+    auto other_geospatial_types = other.bounder_.GeometryTypes();
+    if (geospatial_types.size() != other_geospatial_types.size()) {
+      return false;
+    }
+
+    for (size_t i = 0; i < geospatial_types.size(); i++) {
+      if (geospatial_types[i] != other_geospatial_types[i]) {
+        return false;
+      }
+    }
+
+    return bounder_.Bounds() == other.bounder_.Bounds();
+  }
+
+  void Merge(const GeospatialStatisticsImpl& other) {
+    if (!is_valid_ || !other.is_valid_) {
+      is_valid_ = false;
+      return;
+    }
+
+    bounder_.ReadBox(other.bounder_.Bounds());
+    bounder_.ReadGeometryTypes(other.bounder_.GeometryTypes());
+  }
+
+  void Update(const ByteArray* values, int64_t num_values, int64_t null_count) {
+    if (!is_valid_) {
+      return;
+    }
+
+    geometry::WKBBuffer buf;
+    for (int64_t i = 0; i < num_values; i++) {
+      const ByteArray& item = values[i];
+      buf.Init(item.ptr, item.len);
+      ::arrow::Status status = bounder_.ReadGeometry(&buf);
+      if (!status.ok()) {
+        is_valid_ = false;
+        return;
+      }
+    }
+  }
+
+  void UpdateSpaced(const ByteArray* values, const uint8_t* valid_bits,
+                    int64_t valid_bits_offset, int64_t num_spaced_values,
+                    int64_t num_values, int64_t null_count) {
+    DCHECK_GT(num_spaced_values, 0);
+
+    if (!is_valid_) {
+      return;
+    }
+
+    geometry::WKBBuffer buf;
+    ::arrow::Status status = ::arrow::internal::VisitSetBitRuns(
+        valid_bits, valid_bits_offset, num_spaced_values,
+        [&](int64_t position, int64_t length) {
+          for (int64_t i = 0; i < length; i++) {
+            ByteArray item = SafeLoad(values + i + position);
+            buf.Init(item.ptr, item.len);
+            ARROW_RETURN_NOT_OK(bounder_.ReadGeometry(&buf));
+          }
+
+          return ::arrow::Status::OK();
+        });
+
+    if (!status.ok()) {
+      is_valid_ = false;
+    }
+  }
+
+  void Update(const ::arrow::Array& values) {
+    if (!is_valid_) {
+      return;
+    }
+
+    const auto& binary_array = static_cast<const ::arrow::BinaryArray&>(values);
```

Comment:
it would be nice to check the type before casting.  What about LargeBinary/StringView types? (I thought stringview had a binary equivelant)?